pallet-parachain-upgrade/src/lib.rs

```diff
@@ -0,0 +1,557 @@
+		}
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// Get validation function parameters.
+	///
+	/// This tries to get them first from a magic storage key which is injected
+	/// by cumulus; if that value is set, we should use it, because it's more
+	/// trustworthy than a block's contents. That said, we also try to extract
+	/// them from this block's extrinsics; cumulus also injects them as an
+	/// inherent into each block, so that they're available during block production.
+	fn validation_function_params() -> ValidationFunctionParams {
+		// this storage value is set by cumulus during block validation
+		storage::unhashed::get(VALIDATION_FUNCTION_PARAMS)
+			// this storage value is set each block by the ProvideInherent
+			// implementation, ensuring that the data is also available
+			// during block production
+			.unwrap_or_else(|| VFPs::get())
+			// in the event that something has gone seriously wrong and neither
+			// of those values are present, I believe that VFPs::get falls back
+			// to the Default::default() value, which should be harmless in that
+			// it prevents users from doing any upgrades during the invalid block.
+	}
+
+	/// Put a new validation function into a particular location where polkadot
+	/// monitors for updates. Calling this function notifies polkadot that a new
+	/// upgrade has been scheduled.
+	fn put_polkadot_code(code: &[u8]) {
+		storage::unhashed::put_raw(NEW_VALIDATION_CODE, code);
+	}
+
+	/// Put a new validation function into a particular location where this
+	/// parachain will execute it on subsequent blocks.
+	fn put_parachain_code(code: &[u8]) {
+		storage::unhashed::put_raw(well_known_keys::CODE, code);
+	}
+
+	/// `true` when a code upgrade is currently legal
+	pub fn can_set_code() -> bool {
+		Self::validation_function_params()
+			.code_upgrade_allowed
+			.is_some()
+	}
+
+	/// The maximum code size permitted, in bytes.
+	pub fn max_code_size() -> u32 {
+		Self::validation_function_params().max_code_size
+	}
+}
+
+#[derive(Encode)]
+pub struct ProvideInherentError{}
+
+impl sp_inherents::IsFatalError for ProvideInherentError {
+	fn is_fatal_error(&self) -> bool {
+		// if you can figure out how to return a variant-free struct...
+		true
+	}
+}
+
+impl<T: Trait> ProvideInherent for Module<T> {
+	type Call = Call<T>;
+	type Error = ProvideInherentError;
+	const INHERENT_IDENTIFIER: InherentIdentifier = INHERENT_IDENTIFIER;
+
+	fn create_inherent(data: &InherentData) -> Option<Self::Call> {
+		let vfp = data.get_data(&INHERENT_IDENTIFIER)
+			.and_then(|r| r.ok_or_else(|| "Validation Function Params inherent data not found".into()))
+			.expect("Gets and decodes vfp inherent data");
+
+		VFPs::set(vfp);
+
+		// not sure what precisely the semantics of a None return here are, as
+		// it's not documented, but if I wrote a function like this, then I'd
+		// interpret a None result as "please don't actually create an inherent
+		// for me this block"
+		None
```

Comment:
Assumption about `create_inherent` semantics for which I could not find documentation.